oml/utils/ddp.py

```diff
@@ -0,0 +1,218 @@
+            self.sampler_readed = list(self.sampler)  # type: ignore
+
+        return self.sampler_readed[item]  # type: ignore
+
+    def __len__(self) -> int:
+        return len(self.sampler)  # type: ignore
+
+
+class DDPSamplerWrapper(DistributedSampler):
+    def __init__(
+        self, sampler: TAllSamplers, shuffle_samples_between_gpus: bool = True, pad_data_to_num_gpus: bool = True
+    ):
+        super().__init__(
+            dataset=_Sampler2Dataset(sampler), shuffle=shuffle_samples_between_gpus, drop_last=not pad_data_to_num_gpus
+        )
+        """
+        Default DistributedSampler allows us to built sampler for dataset in DDP mode. Usually we can easily replace
+        default SequentialSampler (when DataLoader(shuffle=False, ...)) and RandomSampler
+        (when DataLoader(shuffle=True, ...)) with DistributedSampler. But for custom sampler we need extra wrapper.
+        With this class we mimic any samplers to dataset and use indices of sampler splitted for each divices.
+        Using this distributed indices we can prepare batch of data.
+        When using DDP we should manage behaviour with last batch, because each device should have the same ammount of
+        data
+        Args:
+            sampler: sequential or batch sampler
+            pad_data_to_num_gpus: When using DDP we should manage behaviour with last batch, because each device should
+                have the same ammount of data. If the sampler length is not evenly divisible by the number of devices,
+                we must duplicate part of the data (pad_data_to_num_gpus=True), or discard part of the
+                data (pad_data_to_num_gpus=False).
+            shuffle_samples_between_gpus: shuffle available indices between feeding to gpu. Note, that shuffle
+            inside gpu after feeding will be used according behaviour of sampler.
+        Note: Wrapper also can be used with default SequentialSampler and RandomSampler, not only custom.
+        """
+
+        self.shift_per_epoch = 0
+        self.sampler = sampler
+
+    def _reload(self) -> None:
+        """
+        We need to reinstantiate wrapper in order to update available indices from sampler for new epoch
+        """
+        self.shift_per_epoch += 1
+        super().__init__(dataset=_Sampler2Dataset(self.sampler), shuffle=self.shuffle, drop_last=self.drop_last)
+        self.set_epoch(self.shift_per_epoch)
+
+    def __iter__(self) -> TAllSamplers:
+        self._reload()
+
+        for sampler_idx in super().__iter__():
+            yield self.dataset[sampler_idx]
+
+
+def patch_dataloader_to_ddp(loader: DataLoader) -> DataLoader:
+    """
+    Function inspects loader and modifies sampler for working in DDP mode.
+    Note:
+        We ALWAYS use padding of samples (number of batches or number of samples per epoch) in order to use same amount
+        of data for each device in DDP, so behaviour with and without DDP may be slightly different (e.g. metrics).
```

Comment:
We ALWAYS use the padding of samples (number of batches or number of samples per epoch) in order to use the same amount
		 of data for each device in DDP, so behavior with and without DDP may be slightly different (e.g. metrics).

Reply:
done